mohssen_fathi_solution/SWAPI.playground/Contents.swift

```diff
@@ -0,0 +1,167 @@
+ - All URLs must be type safe.
+ - All Dates must be type safe.
+
+ ### Bonus Requirements
+
+ Feel free to extend you example with any or all of the following for internet points:
+
+ - Make genders type safe.
+ - Make colors type safe.
+
+ */
+
+struct SWCharacter: Codable {
+
+    let name, height, mass, birthYear: String
+    let gender: Gender
+    let hairColor, skinColor, eyeColor: SWColor
+    let url, homeworld: URL
+    let films, species, vehicles, starships: [URL]
+    let created, edited: Date
+    
+    enum Gender: String, Codable {
+        case male
+        case female
+    }
+    
+    /// Alternatively, could make *Color private
+    enum SWColor: String, Codable {
+        // Hair
+        case blond
+        
+        // Skin
+        case fair
+        
+        // Eye
+        case blue
+        
+        // Could add a computed var here to return UIColor
+    }
+}
+
+extension SWCharacter {
+    static func ==(_ lhs: SWCharacter, _ rhs: SWCharacter) -> Bool {
+        return (
+            lhs.url == rhs.url &&
+            lhs.homeworld == rhs.homeworld &&
+            lhs.name == rhs.name &&
+            lhs.height == rhs.height &&
+            lhs.mass == rhs.mass &&
+            lhs.eyeColor == rhs.eyeColor &&
+            lhs.hairColor == rhs.hairColor &&
+            lhs.skinColor == rhs.skinColor &&
+            lhs.birthYear == rhs.birthYear &&
+            lhs.gender == rhs.gender &&
+            lhs.films == rhs.films &&
+            lhs.species == rhs.species &&
+            lhs.vehicles == rhs.vehicles &&
+            lhs.starships == rhs.starships &&
+            lhs.created == rhs.created &&
+            lhs.edited == rhs.edited
+        )
+    }
+}
+
+
+
+//: A helper function that decodes a `SWCharacter` from raw `Data`.
+//: > Hint: You may need to edit this function. 😉
+func decodeCharacter(from data: Data) -> SWCharacter? {
+    do {
+        let decoder = JSONDecoder()
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+        decoder.dateDecodingStrategy = .formatted(DateFormatter.iso8601Full)
+        let character = try decoder.decode(SWCharacter.self, from: data)
+        return character
+    }
+    catch {
+        print("Error Decoding: \(error.localizedDescription)")
+        return nil
+    }
+}
+
+
+//: A helper function that encodes a `SWCharacter` to raw `Data`.
+//: > Hint: You may need to edit this function. 😉
+func encodeCharacter(_ character: SWCharacter) -> Data? {
+    do {
+        let encoder = JSONEncoder()
+        encoder.keyEncodingStrategy = .convertToSnakeCase
+        encoder.dateEncodingStrategy = .formatted(DateFormatter.iso8601Full)
+        let characterData = try encoder.encode(character)
+        return characterData
+    }
+    catch {
+        print("Error Encoding: \(error.localizedDescription)")
+        return nil
+    }
+}
+
+
+extension DateFormatter {
+    
+    static let iso8601Full: DateFormatter = {
```

Comment:
I like that you made this an extension on `DateFormatter`. I want to go back and change mine after seeing this!

Reply:
👌